ingestion/src/legacy/checked.rs

```diff
@@ -0,0 +1,364 @@
+    #[error("database pool could not return a connection: {0}")]
+    Pool(#[from] bb8::RunError<tokio_postgres::Error>),
+    #[error("error while deserializing message: {0}")]
+    Deserialize(#[from] quick_xml::DeError),
+    #[error(transparent)]
+    Common(#[from] common::Error),
+}
+
+type CheckedMsg = String;
+
+fn parse_message(xmlmsg: &str) -> Result<Vec<UnlabelledDatum>, Error> {
+    // do some checking / further processing of message
+    if !xmlmsg.starts_with("<?xml") {
+        return Err(Error::IssueParsingXML(
+            "kv2kvdata must be xml starting with '<?xml'".to_string(),
+        ));
+    }
+
+    let xmlmsg = match xmlmsg.find("?>") {
+        Some(loc) => &xmlmsg[(loc + 2)..],
+        None => {
+            return Err(Error::IssueParsingXML(
+                "couldn't find end of xml tag '?>'".to_string(),
+            ))
+        }
+    };
+    let item: KvalobsData = quick_xml::de::from_str(xmlmsg)?;
+
+    let mut data: Vec<UnlabelledDatum> = Vec::new();
+
+    // get the useful stuff out of this struct
+    for station in item.stations {
+        for typeid in station.typeids {
+            for obstime in typeid.obstimes {
+                let obs_time =
+                    match NaiveDateTime::parse_from_str(&obstime.val, "%Y-%m-%d %H:%M:%S") {
+                        Ok(time) => time.and_utc(),
+                        Err(e) => {
+                            metrics::counter!(KAFKA_CHECKED_FAILURES).increment(1);
+                            error!(
+                                "time parsing failed in kafka message: {}, original: {}",
+                                Error::IssueParsingTime(e),
+                                &obstime.val
+                            );
+                            continue;
+                        }
+                    };
+                for tbtime in obstime.tbtimes {
+                    // NOTE: tbtime is "table time" which can vary from the actual observation time,
+                    // it's the first time it entered the db in kvalobs. Currently not using it
+                    // TODO: Do we want to handle text data at all? It doesn't seem to be QCed
+                    // if let Some(textdata) = tbtime.kvtextdata {...}
+                    for sensor in tbtime.sensors {
+                        for level in sensor.levels {
+                            if let Some(kvdata) = level.kvdata {
+                                for kvdatum in kvdata {
+                                    data.push(UnlabelledDatum {
+                                        kvid: KvalobsId {
+                                            station: station.val,
+                                            paramid: kvdatum.paramid,
+                                            typeid: typeid.val,
+                                            sensor: sensor.val.unwrap_or(0),
+                                            level: level.val.unwrap_or(0),
```

Comment:
While it makes some sense to default to sensor 0 (its not without its problems also, since could end up with many timeseries all with sensor 0 which could cause confusion), could it make more sense to have a null level? should that be allowed?